compiler/rustc_hir_typeck/src/callee.rs

```diff
@@ -724,6 +729,137 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         err.emit();
     }
 
+    /// Enforces that functions being called actually have the effects that the current context
+    /// requires for them being callable.
+    pub(super) fn enforce_context_effects(&self, hir_id: hir::HirId, span: Span, value: Ty<'tcx>) {
+        if !self.tcx.effects() || self.tcx.sess.opts.unstable_opts.unleash_the_miri_inside_of_you {
+            return;
+        }
+        if value.references_error() {
+            return;
+        }
+
+        // Compute the constness required by the context.
+        let context = self.tcx.hir().enclosing_body_owner(hir_id);
+        if self.tcx.has_attr(context.to_def_id(), sym::rustc_do_not_const_check) {
+            return;
+        }
+        let kind = self.tcx.def_kind(context.to_def_id());
+        let require_const = matches!(
+            kind,
+            DefKind::Const
+                | DefKind::Static(_)
+                | DefKind::AssocConst
+                | DefKind::AnonConst
+                | DefKind::InlineConst
+        );
+        debug_assert_ne!(kind, DefKind::ConstParam);
+
+        let ty::FnDef(did, substs) = *value.kind() else {
+            if require_const {
+                let context = kind.descr(context.to_def_id());
+                self.tcx.sess.span_err(span, &format!("cannot call non-const fn `{value}` in {context}s"));
+            }
+            return;
+        };
+
+        // Tuple struct/variant constructors are not really function calls, they are always fine.
+        if let DefKind::Ctor(_, def::CtorKind::Fn) = self.tcx.def_kind(did) {
+            return;
+        }
+        // FIXME: should const intrinsics also have a constness effect?
+        if self.tcx.is_intrinsic(did) {
+            if self.tcx.is_const_fn(did) {
+                return;
+            }
+        }
+
+        // These special functions do not have effects but are `const fn` because the const evaluator
+        // has logic to just not call them and run special logic instead.
+        if self.tcx.has_attr(did, sym::rustc_do_not_const_check) {
+            return;
+        }
+
+        // Walk all the effects in the type, error on mismatches and bind inferred effects.
+        struct EffectVisitor<'a, 'tcx> {
```

Comment:
:thinking: ideally we check `value: ~const FnOnce<?0>`, think we already talked about that but forgot about the results :grin: 

Can you add a FIXME to change this?

Reply:
oh it's even worse. unsafe functions don't implement any traits... so... I made this horrible commit: https://github.com/rust-lang/rust/pull/101900/commits/82302081c94b793975417df4df7c2bdd2336bfa7

I can upstream it and probably simplify some existing unsafe or non-Rust ABI fn checks